controllers/openshift_controller.go

```diff
@@ -850,72 +850,74 @@ func (r *KataConfigOpenShiftReconciler) processKataConfigDeleteRequest() (ctrl.R
 			r.Log.Error(err, "Error found deleting machine config. If the machine config exists after installation it can be safely deleted manually.",
 				"mc", mc.Name)
 		}
-		// Sleep for MCP to reflect the changes
-		r.Log.Info("Pausing for a minute to make sure mcp has started syncing up")
-		time.Sleep(60 * time.Second)
+		r.Log.Info("Starting to wait for MCO to start")
+		r.kataConfig.Status.WaitingForMcoToStart = true
+		r.kataConfig.Status.UnInstallationStatus.InProgress.IsInProgress = corev1.ConditionTrue
 	}
 
-	mcp := &mcfgv1.MachineConfigPool{}
-	err = r.Client.Get(context.TODO(), types.NamespacedName{Name: machinePool}, mcp)
-	if err != nil {
-		r.Log.Error(err, "Unable to get MachineConfigPool ", "machinePool", machinePool)
-		return ctrl.Result{}, err
+	isConvergedCluster, _ := r.checkConvergedCluster()
+
+	// When nodes migrate from a source pool to a target pool the source
+	// pool is drained immediately and the nodes then slowly join the target
+	// pool.  Thus the operation duration is dominated by the target pool
+	// part and the target pool is what we need to watch to find out when
+	// the operation is finished.  When uninstalling kata on a regular
+	// cluster nodes leave "kata-oc" to rejoin "worker" so "worker" is our
+	// target pool.  On a converged cluster, nodes leave "master" to rejoin
+	// it so "master" is both source and target in this case.
+	targetPool := "worker"
+	if isConvergedCluster {
+		targetPool = "master"
 	}
-	r.Log.Info("Monitoring mcp", "mcp name", mcp.Name, "ready machines", mcp.Status.ReadyMachineCount,
-		"total machines", mcp.Status.MachineCount)
-	r.kataConfig.Status.UnInstallationStatus.InProgress.IsInProgress = corev1.ConditionTrue
+	isMcoUpdating := r.isMcpUpdating(targetPool)
+
+	if !isMcoUpdating && r.kataConfig.Status.WaitingForMcoToStart == true {
+		r.Log.Info("Waiting for MCO to start updating.")
+		// We don't requeue, an MCP going Updated->Updating will
+		// trigger reconciliation by itself thanks to our watching MCPs.
+		return reconcile.Result{}, nil
+	} else {
+		r.Log.Info("No need to wait for MCO to start updating.", "isMcoUpdating", isMcoUpdating, "Status.WaitingForMcoToStart", r.kataConfig.Status.WaitingForMcoToStart)
+		r.kataConfig.Status.WaitingForMcoToStart = false
+	}
+
 	r.clearUninstallStatus()
 	_, result, err2, done := r.updateStatus(machinePool)
 	if !done {
 		return result, err2
 	}
+	err = r.Client.Status().Update(context.TODO(), r.kataConfig)
+	if err != nil {
+		r.Log.Error(err, "Unable to update KataConfig status")
```

Comment:
Do you need to handle the `Conflict` error separately and requeue after a timeout ?

Reply:
That's right and this code is cleaned up in the penultimate commit of this series by removing it altogether.

(Even though it's presented as an addition by diff it's actually not new code, it's old code that wasn't even touched, it just appears moved to the diff algorithm.)